src/kernel/lion_step/cpu.cc

```diff
@@ -0,0 +1,124 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science 
+ *                              and Technology (Skoltech), Russia.
+ *                 2023-present Artificial Intelligence Research 
+ *                              Institute (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is a software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on the StarPU runtime system.
+ *
+ * @file src/kernel/lion_step/cpu.cc
+ * Fused Lion step on buffers on CPU
+ *
+ * @version 1.1.0
+ */
+
+#include "nntile/kernel/lion_step/cpu.hh"
+#include <cmath>             // for std::fabs
+#include "nntile/kernel/cpu.hh" // hypothetical CPU helper
+
+namespace nntile::kernel::lion_step
+{
+
+template<typename T>
+void cpu(Index num_iter, Index num_elems,
+         Scalar beta_1_, Scalar beta_2_,
+         Scalar lambda_, Scalar lr_, Scalar weight_decay_,
+         const T *grad,   // gradient buffer
+         T *first_moment, // buffer for EMA of gradients (m)
+         T *p)            // parameter buffer (theta)
+    noexcept
+//! Fused Lion step on buffers
+/*!
+ * Implements the Lion optimizer step in-place:
+ *
+ *   c_t = beta1 * m_{t-1} + (1 - beta1) * grad
+ *   theta_t = theta_{t-1} - lr * [ sign(c_t) + lambda * theta_{t-1} ]
+ *   m_t = beta2 * m_{t-1} + (1 - beta2) * grad
+ *
+ * Optionally includes weight_decay by adding weight_decay * p_val to grad_val.
+ *
+ * @param[in]  num_iter: Current iteration number (not used directly by Lion, but kept for API symmetry)
+ * @param[in]  num_elems: Number of elements in buffers
+ * @param[in]  beta_1_, beta_2_: Momentum factors
+ * @param[in]  lambda_: Coefficient for parameter penalty in the update
+ * @param[in]  lr_: Learning rate
+ * @param[in]  weight_decay_: L2 regularization coefficient
+ * @param[in]  grad: Input buffer for gradients
+ * @param[inout] first_moment: Stores the moving average of gradients (m)
+ * @param[inout] p: Model parameters updated in-place
+ */
+{
+    using Y = typename T::repr_t;  // e.g. float or double at runtime
+
+    // Convert scalars to underlying type for faster arithmetic
+    const Y beta_1      = static_cast<Y>(beta_1_);
+    const Y beta_2      = static_cast<Y>(beta_2_);
+    const Y lambda_val  = static_cast<Y>(lambda_);
+    const Y lr          = static_cast<Y>(lr_);
+    const Y wd          = static_cast<Y>(weight_decay_);
```

Comment:
Parameter lambda is the weight decay, no need to pass it twice. I see, that you did not understand logic of weight decay correctly, that is why you have 2 of them.